kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/ChecksumLogReplayMetricsTestBase.scala

```diff
@@ -0,0 +1,189 @@
+ */
+
+package io.delta.kernel.defaults
+import java.io.File
+import java.nio.file.Files
+
+import io.delta.kernel.Table
+import io.delta.kernel.internal.fs.Path
+import io.delta.kernel.internal.util.FileNames
+
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+
+/**
+ * Base trait for testing log replay optimizations when reading tables with checksum files.
+ * This trait contains common test setup and test cases but allows specific metadata types
+ * to customize how they load and verify data.
+ *
+ * Test subclasses implement specialized behavior for loading different items:
+ * - PandMCheckSumLogReplayMetricsSuite - tests Protocol and Metadata loading
+ * - DomainMetadataCheckSumReplayMetricsSuite - tests Domain Metadata loading
+ */
+trait ChecksumLogReplayMetricsTestBase extends LogReplayBaseSuite {
+
+  /////////////////////////
+  // Test Helper Methods //
+  /////////////////////////
+
+  /**
+   * Creates a test table with checksum files.
+   * Produces a table with versions 0 to 11 including .json files, .crc files,
+   * and a checkpoint at version 10.
+   *
+   * @param path Path where the table should be created
+   */
+  def buildTableWithCrc(path: String): Unit = {
+    withSQLConf(DeltaSQLConf.DELTA_WRITE_CHECKSUM_ENABLED.key -> "true") {
+      spark.sql(
+        s"CREATE TABLE delta.`$path` USING DELTA AS " +
+          s"SELECT 0L as id")
+      for (_ <- 0 to 10) { appendCommit(path) }
+    }
+  }
+
+  // Abstract method to be implemented by concrete test classes
+  protected def loadSnapshotFieldsCheckMetrics(
+      table: Table,
+      engine: MetricsEngine,
+      expJsonVersionsRead: Seq[Long],
+      expParquetVersionsRead: Seq[Long],
+      expParquetReadSetSizes: Seq[Long],
+      expChecksumReadSet: Seq[Long],
+      version: Long = -1): Unit
+
+  // Method to adjust parquet read set sizes for different implementations
+  protected def getExpectedParquetReadSetSizes(sizes: Seq[Long]): Seq[Long] = sizes
+
+  ///////////
+  // Tests //
+  ///////////
+
+  Seq(-1L, 0L, 3L, 4L).foreach { version => // -1 means latest version
+    test(s"checksum found at the read version: ${if (version == -1) "latest" else version}") {
+      withTempDirAndMetricsEngine { (path, engine) =>
+        // Produce a test table with 0 to 11 .json, 0 to 11.crc, 10.checkpoint.parquet
+        buildTableWithCrc(path)
+        val table = Table.forPath(engine, path)
+
+        loadSnapshotFieldsCheckMetrics(
+          table,
+          engine,
+          // shouldn't need to read commit or checkpoint files as P&M are found through checksum
+          expJsonVersionsRead = Nil,
+          expParquetVersionsRead = Nil,
+          expParquetReadSetSizes = Nil,
+          expChecksumReadSet = Seq(if (version == -1) 11 else version),
+          version)
+      }
+    }
+  }
+
+  test(
+    "checksum not found at read version and checkpoint exists at read version => use checkpoint") {
+    withTempDirAndMetricsEngine { (path, engine) =>
```

Comment:
given all tests require `buildTableWithCrc`, extend `withTempDirAndMetricsEngine` to generate it or add it in beforeTest